James 1988 Add custom jars to Guice+Docker

chibenwa opened the pull request:
Tested by booting JPA and Cassandra with and without additional jars on the mounted volume, and also without the mounted volume.

mbaechler commented on dockerfiles/run/guice/cassandra-ldap/Dockerfile:
```diff
@@ -27,5 +27,10 @@ ADD destination/conf /root/conf
 
-ENTRYPOINT java -Dlogback.configurationFile=/root/conf/logback.xml -Dworking.directory=/root/ -jar james-server.jar 
+ENTRYPOINT java \
+  -classpath '/root/james-server.jar:/root/james-server-cassandra-ldap-guice.lib/*:/root/included-jars/*' \
```
so basically, you broke the jar manifest to be able to add jars at runtime.

chibenwa replied:
1- It was already broken for JPA
2- Isn't the manifest file insite a JAR ?

mbaechler replied:
what I mean is that you are repeating what is inside the manifest, thus, the manifest becomes useless.
I think it's a regression. Maybe it's the price to pay (I doubt), but then we would have to remove the code to generate this manifest.

chibenwa commented:
Yes, I was having a look at this...

chibenwa commented:
I added 

```
                                      <manifestEntries>
                                          <Class-Path>included-jars/*</Class-Path>
                                      </manifestEntries>
```

In the input of the maven-jar-plugin

I checked the generated manifest file, that contains it well:

```
Class-Path: included-jars/* james-server-cassandra-guice.lib/HdrHistog
 ram-2.1.4.jar james-server-cassandra-guice.lib/apache-james-backends-
 cassandra-3.0.0-beta6-SNAPSHOT-tests.jar james-server-cassandra-guice
...
```

Reverted the changes in the dockerfile, but I don't succeed to use the content of the jar

I also tried with :

```
                                      <manifestEntries>
                                          <Class-Path>included-jars/</Class-Path>
                                      </manifestEntries>
```

mbaechler commented:
I guess we have to read the documentation of what a MANIFEST file can  do

chibenwa commented:
I'm reading this. Basically it looks like a folder represents an extracted JAR

chibenwa commented:
Oracle do not seem to document it: https://docs.oracle.com/javase/tutorial/deployment/jar/manifestindex.html

rouazana commented:
If you remove the manifest I think you cannot anymore launch James just doing java -jar james.jar.

chibenwa commented:
Ok, then keep the manifest, and override it (what was originally done) looks like the best option to me

mbaechler commented on server/container/guice/mailet/src/main/java/org/apache/james/utils/GuiceGenericLoader.java:
```diff
@@ -19,24 +19,77 @@
 
 public class GuiceGenericLoader<T> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(GuiceGenericLoader.class);
+    public static final String INCLUDED_JARS_FOLDER_NAME = "included-jars/";
```
what about "extension" ?

chibenwa replied:
you mean naming this folder **extensions-jars** ?

chibenwa replied:
We can realy on streams toArray, which involves a cast (the size of the array being unknown)

mbaechler commented on server/container/guice/mailet/src/main/java/org/apache/james/utils/GuiceGenericLoader.java:
```diff
@@ -19,24 +19,77 @@
+            .flatMap(Throwing.function(this::expendFile).sneakyThrow());
+    }
+
+    private Stream<URL> expendFile(File file) throws MalformedURLException {
```
do you mean expand ?

mbaechler commented on server/container/guice/mailet/src/main/java/org/apache/james/utils/GuiceGenericLoader.java:
```diff
@@ -19,24 +19,77 @@
+            File file = fileSystem.getFile("file://" + INCLUDED_JARS_FOLDER_NAME);
+            ImmutableList<URL> urls = recursiveExpend(file)
+                .collect(Guavate.toImmutableList());
+            return urls.toArray(new URL[urls.size()]);
```
what is the point of collecting to an immutable list if you don't use the list ?

chibenwa replied:
You know a way of collecting to array ?

chibenwa replied:
We can rely on streams toArray, which involves a cast (the size of the array being unknown)

rouazana replied:
stream.toArray?

rouazana replied:
you don't need to know the size: use .toArray(size -> new URL[size])
or I prefer: .toArray(URL[]::new)

aduprat commented on server/container/guice/mailet/src/main/java/org/apache/james/utils/GuiceGenericLoader.java:
```diff
@@ -19,24 +19,77 @@
 
     private final Injector injector;
     private final String defaultPackageName;
+    private URLClassLoader urlClassLoader;
```
final

rouazana commented on server/container/guice/mailet/src/test/java/org/apache/james/utils/GuilceMailetLoaderTest.java:
```diff
@@ -0,0 +1,163 @@
+        GuiceMailetLoader guiceMailetLoader = new GuiceMailetLoader(Guice.createInjector(), RECURSIVE_CLASSPATH_FILE_SYSTEM);
+
+        Mailet mailet = guiceMailetLoader.getMailet(FakeMailetConfig.builder()
+            .mailetName("org.apache.transport.mailets.AIMailet")
```
missing file?

rouazana replied:
can you add a test with a mailet in another package such as com.example.mailets.CustomMailet?

chibenwa replied:
:-( demands me to rebuild this project I threw away....

rouazana commented on server/container/guice/mailet/src/main/java/org/apache/james/utils/GuiceGenericLoader.java:
```diff
@@ -55,25 +53,24 @@ public GuiceGenericLoader(Injector injector, FileSystem fileSystem, String defau
-                .collect(Guavate.toImmutableList());
-            return urls.toArray(new URL[urls.size()]);
+            return (URL[]) recursiveExpand(file)
+                .toArray();
```
.toArray(URL[]::new) to avoid the explicit transtype

rouazana commented on server/container/guice/mailet/src/main/java/org/apache/james/utils/ExtendedClassLoader.java:
```diff
@@ -0,0 +1,82 @@
+        this.urlClassLoader = new URLClassLoader(retrieveIncludedUrls(fileSystem), getClass().getClassLoader());
+    }
+
+    private URL[] retrieveIncludedUrls(FileSystem fileSystem) {
```
ExtensionsUrls?

rouazana commented on server/container/guice/mailet/src/test/java/org/apache/james/utils/GuilceMailetLoaderTest.java:
```diff
@@ -124,38 +71,41 @@ public void getMailetShouldThrowOnBadType() throws Exception {
+            .mailetName("org.apache.james.transport.mailets.Unknown")
             .mailetContext(FakeMailContext.defaultContext())
             .build());
     }
```
Can you please also test with a mailet in an other package than org.apache.james?

chibenwa commented:
test this please

aduprat commented:
test this please

chibenwa commented:
test this please

chibenwa commented:
Merged
